### PR TITLE
Vlad task2

### PR DESCRIPTION
**Pull request by vlad-iancu**

### CODE REVIEW

File: vlad_task2/src/DynamicArray.cpp [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+DynamicArray::DynamicArray()
+{
+    m_dynamicArray = nullptr;
+    m_size = 0;
+}
+
+DynamicArray::DynamicArray(size_t _size)
+{
+    m_dynamicArray = new std::string[_size];
```
Comment (maxlig79) [inline vlad_task2/src/DynamicArray.cpp]: Please use intializer list (in general better performance)

File: vlad_task2/src/DynamicArray.cpp [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+{
+    m_dynamicArray = std::move(source.m_dynamicArray);
+    m_size = std::move(source.m_size);
+}
+
+DynamicArray &DynamicArray::operator=(const DynamicArray &other)
+{
+    if (&other == this)
+    {
+        return *this;
```
Comment (maxlig79) [inline vlad_task2/src/DynamicArray.cpp]: Please avoid multiple return in methods

File: vlad_task2/src/DynamicArray.cpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    delete[] m_dynamicArray;
+    m_size = other.m_size;
+    m_dynamicArray = new std::string[m_size];
+    std::copy_n(other.m_dynamicArray, m_size, m_dynamicArray);
+    return *this;
+}
+
+DynamicArray &DynamicArray::operator=(DynamicArray &&source)
+{
+    delete[] m_dynamicArray;
```
Comment (maxlig79) [inline vlad_task2/src/DynamicArray.cpp]: This will not work for self assignment

File: vlad_task2/src/DynamicArray.cpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    m_dynamicArray = std::move(source.m_dynamicArray);
+    m_size = std::move(source.m_size);
+    return *this;
+}
+
+std::string &DynamicArray::operator[](const size_t index)
+{
+    return m_dynamicArray[index];
```
Comment (maxlig79) [inline vlad_task2/src/DynamicArray.cpp]: You don't need this because you have getEntry

File: vlad_task2/src/DynamicArray.cpp [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+const std::string &DynamicArray::operator[](const size_t index) const
+{
```
Comment (maxlig79) [inline vlad_task2/src/DynamicArray.cpp]: Not needed

File: vlad_task2/src/DynamicArray.cpp [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    return m_dynamicArray[index];
+}
+
+void DynamicArray::deleteEntry(const std::string &entry)
```
Comment (maxlig79) [inline vlad_task2/src/DynamicArray.cpp]: The deleteEntry should return a bool that is true if the element was found and removed.

File: vlad_task2/src/DynamicArray.cpp [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+{
+    std::string *out = new std::string[m_size];
```
Comment (maxlig79) [inline vlad_task2/src/DynamicArray.cpp]: If the entry is in the array you have allocated one extra string

File: vlad_task2/src/DynamicArray.cpp [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    size_t newSize = 0;
+    std::copy_if(m_dynamicArray, m_dynamicArray + m_size, out, [&](const std::string el) {return el != entry; newSize += el != entry;});
+    m_size = newSize;
```
Comment (maxlig79) [inline vlad_task2/src/DynamicArray.cpp]: If the entry is in the array you will have a wrong size for the array.

File: vlad_task2/src/DynamicArray.cpp [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    delete[] m_dynamicArray;
+    m_dynamicArray = out;
+}
+
+std::string DynamicArray::getEntry(const size_t index) const
+{
+    return m_dynamicArray[index];
```
Comment (maxlig79) [inline vlad_task2/src/DynamicArray.cpp:74]: What happen if the index is not in range?

File: vlad_task2/include/DynamicArray.hpp [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+#include <string>
+
+class DynamicArray
+{
+public:
+    DynamicArray();
+    DynamicArray(size_t _size);
+    DynamicArray(const DynamicArray &other);
+    DynamicArray(DynamicArray &&source);
```
Comment (maxlig79) [inline vlad_task2/include/DynamicArray.hpp:11]: If you have implemented this you should use it

File: vlad_task2/include/DynamicArray.hpp [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    DynamicArray &operator=(const DynamicArray &other);
+    DynamicArray &operator=(DynamicArray &&source);
```
Comment (maxlig79) [inline vlad_task2/include/DynamicArray.hpp:13]: If you have implemented this you should use it

File: vlad_task2/SConscript [hunk 12]
```diff
@@ -0,0 +1,25 @@
+from subprocess import check_output
```
Comment (maxlig79) [inline vlad_task2/SConscript:1]: For this task I can't see the benefits of having booth a test and a main,
I suggest you to use only the test(with boost test framework).

File: vlad_task2/src/DynamicArray.cpp [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+void DynamicArray::addEntry(const std::string &entry)
+{
+    reallocate(m_size + 1);
```
Comment (maxlig79) [inline vlad_task2/src/DynamicArray.cpp]: Do you need a parameter for this function? Can it be reallocate()?

**Review by maxlig79 — CHANGES_REQUESTED** [review]

File: vlad_task2/src/DynamicArray.cpp (hunk 8 above)
Comment (nbiro) [reply vlad_task2/src/DynamicArray.cpp]: m_size is always 0 after this function is called...

**Review by maxlig79 — APPROVED** [review]

**Review by ArdeleanuStefan — APPROVED** [review]

**Pull request merged**